### PR TITLE
Audit with paper_trail.

### PR DESCRIPTION
**Pull request by josephbridgwaterrowe**
Add auditing of Entity and Location changes with paper_trail gem.

### CODE REVIEW

File: app/controllers/entities/versions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    def find_entity
+      @entity = Entity.find(params[:id])
+    end
+
+    def hashed_entity_versions
+      @entity
+        .versions
+        .drop(1)
+        .each_with_object({}) do |v, memo|
+        memo[v] = v.reify.decorate
```
Comment (michaelachrisco) [inline app/controllers/entities/versions_controller.rb]:
```Ruby
        .each_with_object({}) { |v, memo| memo[v] = v.reify.decorate }
```

File: config/routes.rb [hunk 2]
```diff
@@ -12,6 +12,7 @@
              :as => 'revoke_role'
     end
   end
```
Comment (michaelachrisco) [inline config/routes.rb]: New line here.

File: spec/features/user_views_previous_entity_version_spec.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
+require 'rails_helper'
+
+feature 'User views previous entity version' do
+  background { sign_in(create(:user, :authenticated, :entity_user)) }
+  given(:entity) do
+    temp_entity = create(:entity)
+    temp_entity = change_entity_name(temp_entity)
+    temp_entity = change_entity_name(temp_entity)
+    temp_entity
```
Comment (michaelachrisco) [inline spec/features/user_views_previous_entity_version_spec.rb]: Is this ```temp_entity``` needed?

File: spec/interactors/create_entity_spec.rb [hunk 4]
```diff
@@ -41,6 +43,20 @@
       its(:persisted?) { is_expected.to be_truthy }
       its(:errors) { is_expected.to be_empty }
       its(:primary_location) { is_expected.to be_present }
+
+      describe 'papertrail' do
+        # Two versions expected due to the primary location update
+        it 'has 2 versions' do
+          expect(entity.versions.size).to eq(2)
+        end
+        # it 'has 1 version' do
```
Comment (michaelachrisco) [inline spec/interactors/create_entity_spec.rb]: remove me.

**Comment by michaelachrisco** [comment]
Other than a few comments, its all good in the hood :+1:

File: spec/features/user_views_previous_entity_version_spec.rb (hunk 3 above)
Comment (josephbridgwaterrowe) [inline spec/features/user_views_previous_entity_version_spec.rb]: No, I suppose I could change this to:

```ruby
given(:entity) do
  temp_entity = create(:entity)
  temp_entity = change_entity_name(temp_entity)
  change_entity_name(temp_entity)
end
```

File: spec/interactors/create_entity_spec.rb (hunk 4 above)
Comment (josephbridgwaterrowe) [inline spec/interactors/create_entity_spec.rb]: Its going to revert back to the 1 version once I move the primary location out.

Comment (michaelachrisco) [inline spec/interactors/create_entity_spec.rb]: Gotcha.

**Comment by josephbridgwaterrowe** [comment]
@michaelachrisco I have made the changes discussed above... please review and merge once the build has completed. Unless I've laid a :hankey: somewhere that I didn't realize of course.

**Comment by michaelachrisco** [comment]
Ill let travis build then merging :)

**Comment by josephbridgwaterrowe** [comment]
Thank you.

Of course this is when you have time, I don't mean to imply that I expect you to be waiting by the merge screen all weekend for it to complete. It can wait until Monday of course. :smiley:

**Pull request merged**